grease/src/Grease/Shape.hs

```diff
@@ -314,3 +338,230 @@ deriving instance (ShowExt ext tag, ShowF tag) => Show (ArgShapes ext tag tys)
   pretty (ArgShapes regs) =
     MC.foldlFC (\doc rShape -> PP.vcat [doc, PP.pretty rShape]) "" regs
+
+-- * Replacing
+
+data TypeMismatch
+  = TypeMismatch
+  { typeMismatchName :: String
+  , expectedType :: C.Some CT.TypeRepr
+  , foundType :: C.Some CT.TypeRepr
+  }
+  deriving Show
+
+instance PP.Pretty TypeMismatch where
+  pretty tm =
+    PP.hsep
+      [ "Type mismatch for"
+      , PP.pretty (typeMismatchName tm) PP.<> ":"
+      , "expected:"
+      , PP.viaShow (expectedType tm)
+      , "but found:"
+      , PP.viaShow (foundType tm)
+      ]
+
+newtype ParsedShapes ext
+  = ParsedShapes {_getParsedShapes :: Map.Map Text (C.Some (Shape ext NoTag))}
+
+-- | Given an initial, provisional list of arguments and a set of replacements
+-- for some of them, calculate a new list of arguments.
+replaceShapes ::
+  forall ext w tys.
+  ExtShape ext ~ PtrShape ext w =>
+  HasPtrWidth w =>
+  -- | Argument names
+  Ctx.Assignment (Const.Const String) tys ->
+  -- | Initial arguments
+  ArgShapes ext NoTag tys ->
+  -- | Replacement arguments
+  ParsedShapes ext ->
+  Either TypeMismatch (ArgShapes ext NoTag tys)
+replaceShapes names (ArgShapes args) (ParsedShapes replacements) =
+  -- TODO: Check that all the map keys are expected
+  ArgShapes
+    <$> Ctx.zipWithM (\(Const.Const nm) s -> replaceOne nm s) names args
+ where
+  replaceOne :: String -> Shape ext NoTag t -> Either TypeMismatch (Shape ext NoTag t)
+  replaceOne nm s =
+    case Map.lookup (Text.pack nm) replacements of
+      Just (C.Some replace) ->
+        let ty = shapeType ptrShapeType s
+         in let ty' = shapeType ptrShapeType replace
+             in case testEquality ty ty' of
+                  Just Refl -> Right replace
+                  Nothing ->
+                    Left $
+                      TypeMismatch
+                        { typeMismatchName = nm
+                        , expectedType = C.Some ty
+                        , foundType = C.Some ty'
+                        }
+      Nothing -> Right s
+
+isInSubProg :: Word64 -> Subprogram -> Bool
+isInSubProg w sub =
+  let entryMatch = (==) w <$> MDwarf.subEntryPC sub
+      def = fromMaybe False
+   in def
+        ( do
+            sdef <- MDwarf.subDef sub
+            lpc <- MDwarf.subLowPC sdef
+            hpc <- MDwarf.subHighPC sdef
+            let res = w >= lpc && w < (lpc + hpc)
+            pure $ res
+        )
+        || def entryMatch
+
+rightToMaybe :: Either a b -> Maybe b
+rightToMaybe (Left _) = Nothing
+rightToMaybe (Right b) = Just b
+
+extractType :: Subprogram -> MDwarf.TypeRef -> Maybe MDwarf.TypeApp
+extractType sprog vrTy =
+  do
+    let mp = MDwarf.subTypeMap sprog
+    (mbtypeApp, _) <- Map.lookup vrTy mp
+    rightToMaybe mbtypeApp
+
+constructPtrTarget :: HasPtrWidth w => Subprogram -> MDwarf.TypeApp -> Maybe (PtrTarget w NoTag)
+constructPtrTarget sprog tyApp =
+  PtrTarget Nothing <$> shapeSeq tyApp
+ where
+  -- TODO: are these always byte fields?
```

Comment:
Not sure I understand the question but let me know if there's anything I can clarify or help with here!